teps/0048-pipeline-as-code.md

```diff
@@ -0,0 +1,214 @@
+
+## Summary
+
+The pipelines as code technique emphasizes that the configuration of delivering
+pipelines that build, test and deploy our applications or infrastructure should
+be treated as code; they should be placed under source control and modularized
+in reusable components with automated testing and deployment.
+
+
+## Motivation
+
+The Tekton pipeline definitions are currently dissociated from a source code repository. 
+They may be located in the same repository as they code they should build, test and deploy,
+but Tekton is not aware of them. Tekton expects the pipeline definition to be deployed to a
+kubernetes cluster. Even when the Tekton resources are deployed to the cluster from the
+Repository, it's not easy to test changes to them, as their definition won't be updated until
+they are redeployed to the cluster. 
+
+The reason for the current behaviour is that Tekton is non-prescriptive about how to set up
+CI and CD pipelines. It aims to provide the building blocks for creating such pipelines and it leaves
+it up to the user to decide how to integrate the pipeline in the overall workflow. 
+
+The use case of a single repo, running CI/CD jobs in isolation, is very common and
+we wish to provide dedicated support for it, by providing an easy way for users to
+run pipelines and tasks stored in a specific folder of the repo. The same tool can be used by
+pipeline and task authors for quick testing and experimentation of their code.
+
+### Goals
+
+A way for users to add a tekton pipeline in a repository and getting it
+automatically picked up for CI or CD.
+
+> The solution should be as much as possible git hosting system agnostic.
+
+### Non-Goals
+
+- GitOps: Pipelines as Code does not take care of the [GitOps primitives](https://www.weave.works/technologies/gitops/), it
+  simply applies tekton pipelines that may take care of some GitOps operation.
+
+### Use Cases
+
+- A User have a source code repository based on some Golang backend.
+- User add a webhook (or other means) from its repository to pipelines as a code.
+- The user add a directory called in `.tekton` with some pipeline definition for
+  testing its repository.
+- User sending a PR with its change.
+- The source code is automatically tested with the pipeline definition in `.tekton`.
+
+Subsequently if the user needs to add another component to its source code (i.e:
+a javascript based frontend) and need to make a changes to the pipeline
+definition to do the testing of that new component, the change to the pipeline
+would only apply for that Pull Request until it will be merged.
+
+## Requirements
+
+* We need to be able to test changes inside the `.tekton` folder as part of the PR sent.
+* We should provide an easy way (single curl command and/or operator flag) to deploy Tekton with "tekton-as-code" support.
+* We need assume single tenancy on the pipelines-as-code cluster.
+* We assume single tenancy on the tekton-as-code cluster
```

Comment:
can you explain a bit more about what assuming single tenancy means here and why it's a requirement?

(is this a duplicate of the line above?)

Reply:
I think it's the *initial* assumption, to simplify the design a bit, @afrittoli @chmouel ?

Reply:
yep 👍🏻 

I wasn't sure about this at first, since multi tenancy is definitively something I would like to have, but there is quite a bit to talk about on multi tenancy support with tekton and it's not necessary a first class citizen on vanilla kubernetes too.